Specification for tail-call elimination and `@tailrec`

som-snytt opened the pull request:
Fixes scala/bug#1219

Mention two mandated optimizations.

som-snytt committed:
Specify tail-recursive call elimination and annotation

lrytz commented:
👍 thanks!

lrytz commented on spec/04-basic-declarations-and-definitions.md:
```diff
@@ -840,6 +840,23 @@ class C extends I {
+before returning from the method are replaced with jumps to the beginning of
+the method, much as in a while loop.
+
+This transform is performed automatically, without intervention by the user,
+but may be enforced using `scala.annotation.tailrec`, for cases where
```
Slightly ambiguous? "may be enforced" could mean "the user can enforce" or "the compiler may enforce".

SethTisue commented:
@sjrd do you think it's appropriate to spec this...?

I'm definitely okay with the chapter 11 addition, but I wonder a bit about chapter 4. I'm aware that the compiler performs the optimization long before bytecode generation, but it still seems arguably odd to me to specify something that's ultimately about the runtime rather than about semantics per se.

som-snytt commented:
I also had a second thought, especially because there are potentially many behaviors with associated annotations.

In this case, however, the user must be able to rely on TCE. It must be considered an idiom of the language. Otherwise, everyone must write while loops because they are portable.

There may be additional constraints to spec, such as tail calls to a different instance.

sjrd commented on spec/11-annotations.md:
```diff
@@ -39,17 +39,28 @@ String @local                               // Type annotation
+  * `@switch` Marks the expression submitted to a match as "switchable",
+    such that the match can be compiled to an efficient form.
+    The compiler will warn if the type of the expression is not a switchable type.
+    Certain degenerate matches may remain unoptimized without a warning.
```
These two are not specific to the Java Platform. They should go in a separate section for platform-agnostic annotations.

sjrd commented on spec/04-basic-declarations-and-definitions.md:
```diff
@@ -840,6 +840,23 @@ class C extends I {
 Here, it is OK to leave out the result type of `factorial`
 in `C`, even though the method is recursive.
 
+### Tail-call Elimination
```
```suggestion
### Tail-Recursive Call Elimination
```
We don't have general tail call elimination

som-snytt replied:
The text says as much, but since hyphenization and capitalization are also sources of controversy, I take the recommended edit without further ado.

sjrd commented:
IMO it makes sense for it to be specified, indeed. But then we should also specify what calls are actually considered tail-recursive, otherwise we still cannot really rely on it, can we?

som-snytt commented:
I changed the order of sections, prioritizing "compiler annotations", then incongruously leaving "deprecation annotations" in their own section (without adding any), then "platform annotations".

I also amended some funky line lengths. Apologies for the resulting diff.

The edits as requested are to tighten up "tail with recursive call, pronounced 'twerk'" (to which I also added "sibling-calls", actually), moving tailrec and switch to compiler annotations.

som-snytt commented:
I still need to add a section for "Non-Sensical Annotations".

SethTisue commented on spec/04-basic-declarations-and-definitions.md:
```diff
@@ -840,6 +840,23 @@ class C extends I {
+Method definitions which contain self-recursive invocations in tail position
+are optimized for stack safety. Self-invocations which are the last operation
+before returning from the method are replaced with jumps to the beginning of
+the method, much as in a while loop. Sibling-invocations are also optimized.
```
I think it needs to be explained what a "sibling-invocation" is.

(Wild guess: is it when `this` changes...? It took me some pondering to even formulate the guess.)

som-snytt replied:
I studied Wikipedia for usages slash nomenclature. https://en.wikipedia.org/wiki/Tail_call

SethTisue commented on spec/04-basic-declarations-and-definitions.md:
precede with `import scala.annotation.tailrec`?

som-snytt replied:
Odersky is on the hot seat to improve that, just because it irks you. I loosely said annotations are found in either package.

SethTisue commented on spec/04-basic-declarations-and-definitions.md:
```diff
@@ -840,6 +840,23 @@ class C extends I {
+the method, much as in a while loop. Sibling-invocations are also optimized.
+
+This transform is performed automatically, without intervention by the user;
+the user can require it with the annotation, `scala.annotation.tailrec`,
```
I think the "require it" wording is insufficiently clear and explicit. Maybe add something like "If the annotation is present but the compiler cannot perform the transform, an error is signaled", that kind of thing?

som-snytt replied:
Nth told me elsewhere, That's not an assert, it's a require. This is `require` or error. However, as this is third edit, I will try to incorporate your words. Oh, maybe "require the transform for successful compilation."

som-snytt replied:
I'm glad I put ci on last-only. This could take a while.

som-snytt commented:
Applied additional verbal screws.

sjrd reviewed (approved):
I still think it would be good to specify what is a tail position. But as is, this PR seem progress to me.

som-snytt commented:
I will volunteer to backport words from the Scala 3 spec.

SethTisue commented:
thx AP!
